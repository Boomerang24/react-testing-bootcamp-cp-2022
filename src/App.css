body {
  background-color: #282c34;
}

.App {
  text-align: center;
  max-width: 95rem;
  margin: auto;
  padding: 0 2em;
}

.App-header {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-header h1 {
  text-decoration: underline;
}

.picture-information-calendar {
  background-color: #42b6d6;
  border-radius: 1.5rem;
  font-size: 2rem;
  padding: 0.5rem 1rem;
}

.main-wrapper {
  width: 100%;
}

.main-sections {
  display: grid;
  grid-template-columns: repeat(1, 1fr);
  text-align: center;
}

.picture-description {
  align-items: flex-end;
  color: rgb(138, 138, 138);
  display: flex;
  font-size: 1rem;
  margin-top: -1rem;
  text-align: start;  
}

.picture-description:hover {
  color: whitesmoke
}

.picture-information {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  /* max-width: 100%; */
}

.picture-information p {
  font-size: 0.4em;
  text-align: end;
  margin: 0;
  padding: 0 0.5em;
}

.picture-container {
  background-color: black;
  border-radius: 7px;
  height: 16em;
  width: 100%;
}

.picture-container img {
  height: 16em;
  object-fit: contain;
}

.mainscreen-footer {
  font-size: 0.9em;
  font-style: italic;
}
