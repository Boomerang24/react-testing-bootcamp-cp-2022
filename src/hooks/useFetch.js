import { useEffect, useState } from 'react';

export const useFetch = (url) => {
  const [response, setResponse] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const doFetch = async (url) => {
      setResponse(null);
      setLoading(true);
      setError(null);
      try {
        const res = await fetch(url);
        const json = await res.json();
        if (!res.ok) {
          setError(json.error);
        } else {
          setResponse(json);
        }
      } catch (error) {
        console.log(error);
      } finally {
        setLoading(false);
      }
    };
    doFetch(url);
  }, [url]);
  return { response, loading, error };
};
