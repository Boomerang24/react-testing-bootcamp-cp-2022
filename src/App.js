/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
/* eslint-disable no-undef */
/* eslint-disable react/no-unescaped-entities */
import { forwardRef, useState } from 'react';
import DatePicker from 'react-datepicker';

import './App.css';
// import 'react-datepicker/dist/react-datepicker.css';

const testImg = 'https://apod.nasa.gov/apod/image/2204/PlanetParadeSydney_Agrawal_960_ann.jpg';

function App() {
  const [startDate, setStartDate] = useState(new Date());
  const ExampleCustomInput = forwardRef(({ value, onClick }, ref) => (
    <button className="example-custom-input" onClick={onClick} ref={ref}>
      {value}
    </button>
  ));
  return (
    <>
      <div className="App">
        <header className="App-header">
          <h1>Picture of the Day</h1>
          <div className="main-wrapper">
            <DatePicker
              className="date-picker"
              selected={startDate}
              onChange={(date) => setStartDate(date)}
              customInput={<ExampleCustomInput />}
            />
            <div className="main-sections">
              <div className="picture-information">
                <h3>Planet Parade over Sydney Opera House</h3>
                <p>2022-04-26</p>
                <div className="picture-container">
                  {/* <img src="/testImg.jpg" alt="img" /> */}
                  <img src={testImg} alt="img" />
                </div>
              </div>
              <div className="picture-description">
                <p>
                  The world is waking up to a picturesque planet parade. Just before dawn, the
                  eastern skies over much of planet Earth are decorated by a notable line of
                  familiar planets. In much of Earth's northern hemisphere, this line of planets
                  appears most nearly horizontal, but in much of Earth's southern hemisphere, the
                  line appears more nearly vertical. Pictured over the Sydney Opera House in
                  southern Australia, the planet line was captured nearly vertical about five days
                  ago. From top to bottom, the morning planets are Saturn, Mars, Venus, and Jupiter.
                  As April ends, the angular distance between Venus and Jupiter will gradually pass
                  below a degree as they switch places. Then, as May ends, Jupiter will pass near
                  Mars as those two planets switch places. In June, the parade will briefly expand
                  to include Mercury. Notable Submissions to APOD: Morning Planet Parade 2022
                </p>
              </div>
            </div>
            <div className="mainscreen-footer">
              <p>Project created during Wizeline Academy React Testing Bootcamp</p>
            </div>
          </div>
        </header>
      </div>
    </>
  );
}

export default App;
