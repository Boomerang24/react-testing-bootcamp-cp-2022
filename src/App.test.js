/* eslint-disable no-undef */
import { fireEvent, render, screen } from '@testing-library/react';
import moment from 'moment';

import { useFetch } from './hooks/useFetch';
import App from './App';

const setup = () => render(<App />);

const failedResponse = {
  code: 'API_KEY_MISSING',
  message: 'No api_key was supplied. Get one at https://api.nasa.gov:443',
};

jest.mock('./hooks/useFetch', () => ({
  useFetch: jest.fn(),
}));

const sampleTodaysInfo = {
  url: 'https://www.youtube.com/embed/m8qvOpcDt1o?rel=0',
  title: 'Animation: Odd Radio Circles',
  explanation: 'Sample Explanation',
  media_type: 'video',
};

/* eslint-disable no-undef */
describe('App Tests', () => {
  it('should render "mainscreen" component correctly', () => {
    const { title } = sampleTodaysInfo;

    useFetch.mockImplementation(() => ({
      response: sampleTodaysInfo,
      loading: false,
    }));
    const AppComponent = setup();
    const { getByRole, getByTitle, getByText } = AppComponent;

    const header = getByRole('heading', { name: title });
    const videoContainer = getByTitle(`Video - ${title}`);
    const description = getByText('Sample Explanation');

    expect(AppComponent).toMatchSnapshot();
    expect(videoContainer).toBeTruthy();
    expect(header).toBeInTheDocument();
    expect(description).toBeInTheDocument();
  });
  it('should render "loading" component correctly', () => {
    useFetch.mockImplementation(() => ({
      response: null,
      loading: true,
    }));
    const AppComponent = setup();
    const { getByTestId } = AppComponent;

    const loader = getByTestId('triangle-loading');

    expect(AppComponent).toMatchSnapshot();
    expect(loader).toBeInTheDocument();
    expect(loader).toHaveClass('screen-loader');
  });
  it('should display "Future dates are not valid"', async () => {
    const tomorrow = moment().add(1, 'days').format('YYYY-MM-DD');
    useFetch.mockImplementation(() => ({
      response: null,
      loading: false,
      error: {
        code: 400,
        msg: `Date must be between Jun 16, 1995 and Apr 29, 2022.`,
        service_version: 'v1',
      },
    }));

    setup();
    const DatePicker = await screen.getByLabelText('calendar-picker');

    fireEvent.change(DatePicker, { target: { value: tomorrow } });

    expect(await screen.findByText('Future dates are not valid')).toBeInTheDocument();
  });

  it('should return "There was an error, please try again" when failing to fetch', async () => {
    useFetch.mockImplementation(() => ({
      response: null,
      loading: false,
      error: failedResponse,
    }));
    setup();

    expect(await screen.findByText('There was an error, please try again')).toBeInTheDocument();
  });
});
