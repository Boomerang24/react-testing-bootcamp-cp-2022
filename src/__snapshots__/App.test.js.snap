// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Tests should render "loading" component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="screen-loader"
        data-testid="triangle-loading"
        role="presentation"
        style="display: flex;"
      >
        <svg
          aria-label="loading"
          class="sc-gKXOVf RDHOs"
          data-testid="triangle-svg"
          height="150"
          id="triangle"
          viewBox="-3 -4 39 39"
          width="150"
        >
          <polygon
            class="sc-jSMfEi ddtSqb"
            fill="transparent"
            points="16,0 32,32 0,32"
            stroke="red"
            stroke-width="1"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="screen-loader"
      data-testid="triangle-loading"
      role="presentation"
      style="display: flex;"
    >
      <svg
        aria-label="loading"
        class="sc-gKXOVf RDHOs"
        data-testid="triangle-svg"
        height="150"
        id="triangle"
        viewBox="-3 -4 39 39"
        width="150"
      >
        <polygon
          class="sc-jSMfEi ddtSqb"
          fill="transparent"
          points="16,0 32,32 0,32"
          stroke="red"
          stroke-width="1"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App Tests should render "mainscreen" component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header
          class="App-header"
        >
          <div
            class="Toastify"
          />
          <h3>
            NASA - Picture of the Day
          </h3>
          <div
            class="main-wrapper"
          >
            <input
              aria-label="calendar-picker"
              class="picture-information-calendar"
              id="calendarPicker"
              max="2022-04-29"
              name="date-picker"
              type="date"
              value="2022-04-29"
            />
            <div
              class="main-sections"
            >
              <div
                class="picture-information"
              >
                <h3>
                  Animation: Odd Radio Circles
                </h3>
                <div
                  class="picture-container"
                >
                  <iframe
                    class="picture"
                    src="https://www.youtube.com/embed/m8qvOpcDt1o?rel=0"
                    title="Video - Animation: Odd Radio Circles"
                  />
                </div>
              </div>
              <div
                class="picture-description"
              >
                <p>
                  Sample Explanation
                </p>
              </div>
            </div>
            <div
              class="mainscreen-footer"
            >
              <p>
                Project created during Wizeline Academy React Testing Bootcamp
              </p>
            </div>
          </div>
        </header>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header
        class="App-header"
      >
        <div
          class="Toastify"
        />
        <h3>
          NASA - Picture of the Day
        </h3>
        <div
          class="main-wrapper"
        >
          <input
            aria-label="calendar-picker"
            class="picture-information-calendar"
            id="calendarPicker"
            max="2022-04-29"
            name="date-picker"
            type="date"
            value="2022-04-29"
          />
          <div
            class="main-sections"
          >
            <div
              class="picture-information"
            >
              <h3>
                Animation: Odd Radio Circles
              </h3>
              <div
                class="picture-container"
              >
                <iframe
                  class="picture"
                  src="https://www.youtube.com/embed/m8qvOpcDt1o?rel=0"
                  title="Video - Animation: Odd Radio Circles"
                />
              </div>
            </div>
            <div
              class="picture-description"
            >
              <p>
                Sample Explanation
              </p>
            </div>
          </div>
          <div
            class="mainscreen-footer"
          >
            <p>
              Project created during Wizeline Academy React Testing Bootcamp
            </p>
          </div>
        </div>
      </header>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
